### PR TITLE
Add custom marshalling and unmarshalling for block Header

### PR DESCRIPTION
**Pull request by 0xAleksaOpacic**
# Description

Polygon Edge Header cannot be unmarshalled with go-etherem ethClient. This PR implements custom json marshal and unrmashal for Header so it is compatible with other EVM networks

# Changes include

- [x] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Breaking changes

I don't think this is a breaking change
I've manually tested:
started a network on the latest develop branch, switching to this branch, adding a new node to bulkSync, waiting in watchSync, voting to become validator, watching consensus 

# Checklist

- [x] I have assigned this PR to myself
- [ ] I have added at least 1 reviewer
- [x] I have added the relevant labels
- [ ] I have updated the official documentation
- [ ] I have added sufficient documentation in code

## Testing

- [x] I have tested this code with the official test suite
- [x] I have tested this code manually

### Manual tests

I've created go-ethereum ethClient script to test this feature. If you run this script on develop branch you will see bunch of errors
```

package main

import (
	"context"
	"fmt"
	"log"

	"github.com/ethereum/go-ethereum/core/types"
	"github.com/ethereum/go-ethereum/ethclient"
)

func main() {
	client, err := ethclient.Dial("ws://localhost:10002/ws")
	if err != nil {
		log.Fatal(err)
	}

	headers := make(chan *types.Header)
	sub, err := client.SubscribeNewHead(context.Background(), headers)
	if err != nil {
		log.Fatal(err)
	}

	for {
		select {
		case err := <-sub.Err():
			log.Fatal(err)
		case header := <-headers:
			fmt.Println(header)

		}
	}
}
```

# Additional comments

Fixes EDGE-440

### CODE REVIEW

File: helper/hex/hex.go [hunk 1]
```diff
@@ -52,6 +52,17 @@ func EncodeUint64(i uint64) string {
 	return string(strconv.AppendUint(enc, i, 16))
 }
 
+// DecodeUint64 decodes a hex string with 0x prefix to uint64
+func DecodeUint64(hexStr string) uint64 {
+	// remove 0x suffix if found in the input string
+	cleaned := strings.Replace(hexStr, "0x", "", -1)
```
Comment (Kourin1996) [inline helper/hex/hex.go]: `TrimPrefix` is better?

File: helper/hex/hex.go [hunk 2]
```diff
@@ -52,6 +52,17 @@ func EncodeUint64(i uint64) string {
...
+
+	// base 16 for hexadecimal
+	result, _ := strconv.ParseUint(cleaned, 16, 64)
```
Comment (Kourin1996) [inline helper/hex/hex.go]: Why not check error?

Comment (0xAleksaOpacic) [reply helper/hex/hex.go]: My mistake we should check the error 
fixed 406fe2c54dacc8d6d08496f8a2829cc85b2e95e1 + ff727c450cd58043583531e0ac4fa9bf96d81d72
I've also added a test to confirm that encode, decode, and MustDecode works, just wasn't sure how to reproduce error. If you have an idea please help :)

File: helper/hex/hex.go (hunk 1 above)
Comment (0xAleksaOpacic) [reply helper/hex/hex.go]: Do you mean TrimPrefix is better than cleaned? 
Maybe slicedHex or formattedHex, trimmedHex? What do you think

Comment (Kourin1996) [reply helper/hex/hex.go]: No, I meant strings package had `TrimPrefix` function

File: helper/hex/hex.go [hunk 3]
```diff
@@ -52,6 +52,35 @@ func EncodeUint64(i uint64) string {
...
+	// remove 0x suffix if found in the input string
+	cleaned := strings.Replace(hexStr, "0x", "", -1)
+
+	// base 16 for hexadecimal
+	result, err := strconv.ParseUint(cleaned, 16, 64)
+	if err != nil {
+		panic(fmt.Errorf("could not decode hex: %w", err))
+	}
+
+	return result
```
Comment (Kourin1996) [inline helper/hex/hex.go]: Why not call DecodeUint64? I recommend you not to implement the same code twice

File: types/header_test.go [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+		t.Fatalf("Unable marshall header,  %v", err)
+	}
+
+	unmarshallHeader := Header{}
+	assert.NoError(t, json.Unmarshal(headerJSON, &unmarshallHeader))
+
+	if !reflect.DeepEqual(unmarshallHeader, header) {
+		t.Fatal("Resulting data and expected values are not equal")
+	}
+}
```
Comment (Kourin1996) [inline types/header_test.go]: Can you add test to check unmarshalling from json?

**Review by Kourin1996 — CHANGES_REQUESTED** [review]

File: helper/hex/hex.go (hunk 1 above)
Comment (Aleksao998) [reply helper/hex/hex.go]: fied 36cd3c09d6ee00272a1ce9454d16bb685e06ce8f

Comment (0xAleksaOpacic) [reply helper/hex/hex.go]: Thanks agree 204fec563dfe954c6af8e159b5f50495fd4631aa

File: helper/hex/hex.go (hunk 3 above)
Comment (0xAleksaOpacic) [reply helper/hex/hex.go]: thanks agree, fixed 204fec563dfe954c6af8e159b5f50495fd4631aa

**Comment by scalalang2** [comment]
related #643

**Comment by scalalang2** [comment]
Another test code here.

```golang
func TestBlockByHash(t *testing.T) {
	edgeNet := "https://rpc-edgenet.polygon.technology"
	// https://explorer-edgenet.polygon.technology/block/1320009/transactions
	blockHash := "0xd4f333b237f56436f6e3074b22a1e0dada6dff4ee574778fa5dcdcebe118323d"
	client, err := ethclient.DialContext(context.Background(), edgeNet)
	if err != nil {
		t.Errorf("failed to connect to the network")
	}

	block, err := client.BlockByHash(context.Background(), ethcommon.HexToHash(blockHash))
	if err != nil {
		t.Errorf("failed to fetch block")
	}

	if block.Hash().Hex() != blockHash {
		t.Errorf("received block is not eqaul, query: %s, got: %s", blockHash, block.Hash().Hex())
	}
}
```

**Comment by 0xAleksaOpacic** [comment]
> related #643
> 
> If this PR is merged, Can I expect it would works well without modifying my codes written with ethClient?

Hi,
I need to investigate what is the cause of this issue. If it is something related with this PR it will be included if not i will open separate PR :)

File: helper/hex/hex.go [hunk 5]
```diff
@@ -52,6 +52,31 @@ func EncodeUint64(i uint64) string {
 	return string(strconv.AppendUint(enc, i, 16))
 }
 
+// DecodeUint64 decodes a hex string with 0x prefix to uint64
+func DecodeUint64(hexStr string) (uint64, error) {
+	// remove 0x suffix if found in the input string
+	cleaned := strings.TrimPrefix(hexStr, "0x")
+
+	// base 16 for hexadecimal
+	result, err := strconv.ParseUint(cleaned, 16, 64)
```
Comment (zivkovicmilos) [inline helper/hex/hex.go]: This can be inlined

File: helper/hex/hex.go [hunk 6]
```diff
@@ -52,6 +52,31 @@ func EncodeUint64(i uint64) string {
...
+	if err != nil {
+		return 0, err
+	}
+
+	return result, nil
+}
+
+// MustDecodeUint64 decodes a hex string with 0x prefix to uint64
+// Returns panic if there is an error
+func MustDecodeUint64(hexStr string) uint64 {
```
Comment (zivkovicmilos) [inline helper/hex/hex.go]: Please avoid adding methods that contain explicit panics

File: helper/hex/hex_test.go [hunk 7]
```diff
@@ -0,0 +1,42 @@
+package hex
+
+import (
+	"github.com/stretchr/testify/assert"
+	"testing"
+)
+
+func TestEncodingUint64(t *testing.T) {
```
Comment (zivkovicmilos) [inline helper/hex/hex_test.go]: Why not have this test simply generate a random number array (with 0 included), and test it out that way?

File: helper/hex/hex_test.go [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+	testTable := []struct {
+		name       string
+		value      uint64
+		shouldFail bool
```
Comment (zivkovicmilos) [inline helper/hex/hex_test.go]: Why have this in the test if none of your test cases fail?

File: helper/hex/hex_test.go [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+			decodedValue, err := DecodeUint64(encodedValue)
+			if err != nil && !testCase.shouldFail {
+				t.Fatalf("Unable to decode uint64, %v", err)
+			}
+
+			assert.Equal(t, testCase.value, decodedValue)
+
+			decodedValue = MustDecodeUint64(encodedValue)
+
+			assert.Equal(t, testCase.value, decodedValue)
```
Comment (zivkovicmilos) [inline helper/hex/hex_test.go]: This is redundant, since `MustDecodeUint64` is redundant

File: types/header.go [hunk 10]
```diff
@@ -77,6 +176,15 @@ func (n Nonce) MarshalText() ([]byte, error) {
 	return []byte(n.String()), nil
 }
 
+func (n *Nonce) UnmarshalText(input []byte) error {
+	input = input[2:]
```
Comment (zivkovicmilos) [inline types/header.go]: What is this magic number `2`?

Comment (zivkovicmilos) [reply types/header.go]: Aha, the hex prefix.

Please make this clear

File: helper/hex/hex.go (hunk 6 above)
Comment (0xAleksaOpacic) [reply helper/hex/hex.go]: I've added this method following the standard from this package. As you can see other decode methods (https://github.com/0xPolygon/polygon-edge/blob/develop/helper/hex/hex.go#L38) have MustDecode versions used mainly for tests. Agree we can remove this only because we don't want to risk that someone will use this i code and trigger panic.
fixed: a50aef5ac33eb44c76d87c8367a20a1158c146fe

File: helper/hex/hex_test.go (hunk 9 above)
Comment (0xAleksaOpacic) [reply helper/hex/hex_test.go]: fixed a50aef5ac33eb44c76d87c8367a20a1158c146fe

File: types/header.go [hunk 11]
```diff
@@ -77,6 +176,15 @@ func (n Nonce) MarshalText() ([]byte, error) {
 	return []byte(n.String()), nil
 }
 
+func (n *Nonce) UnmarshalText(input []byte) error {
```
Comment (zivkovicmilos) [inline types/header.go:173]: Why does this method assume that `Nonce` will be hex encoded?

File: types/header.go [hunk 12]
```diff
@@ -30,6 +32,103 @@ type Header struct {
 	Hash         Hash    `json:"hash"`
 }
 
+// HeaderJson represents a block header used for json calls
+type HeaderJSON struct {
```
Comment (zivkovicmilos) [inline types/header.go]: Why does this need to be a separate struct for JSON?

Comment (zivkovicmilos) [reply types/header.go]: Why not define a simple JSON marshal method for the existing struct?

Comment (zivkovicmilos) [reply types/header.go]: I see that you use it in the marshal method for the old `Header` type.

Why not define it locally, why is it exported?

**Review by zivkovicmilos — CHANGES_REQUESTED** [review]
Left some comments that need resolving 🙏 

Please look at @Kourin1996's comments as well.

Overall it looks good 👍

File: helper/hex/hex.go (hunk 5 above)
Comment (0xAleksaOpacic) [reply helper/hex/hex.go]: fixed 21948606924f88571f8f05e9df02242792ceac8c

File: types/header.go (hunk 10 above)
Comment (0xAleksaOpacic) [reply types/header.go]: Fixed 06d747cdb9a1b7d260f36baf70ae71ce32455c9d. Do you think we should use this PR to change in other packages as well

Comment (zivkovicmilos) [reply types/header.go]: Why is this an exported method?

Comment (0xAleksaOpacic) [reply types/header.go]: Because every unmarshal method in our code is exported. I don't see a reason why it shouldn't be

Comment (0xAleksaOpacic) [reply types/header.go]: Also i am positive that marshaling and unmarshaling methods needs to be exported

File: helper/hex/hex_test.go (hunk 7 above)
Comment (0xAleksaOpacic) [reply helper/hex/hex_test.go]: fixed 8ef3965ab7c423466a31342f16a9715bd53826b6

File: helper/hex/hex_test.go (hunk 8 above)
Comment (0xAleksaOpacic) [reply helper/hex/hex_test.go]: fixed 8ef3965ab7c423466a31342f16a9715bd53826b6

Comment (0xAleksaOpacic) [reply helper/hex/hex_test.go]: fixed lint 70e04dfc8686d02178eb0868d3bf78e3f7d54f4f

File: helper/hex/hex_test.go (hunk 7 above)
Comment (0xAleksaOpacic) [reply helper/hex/hex_test.go]: fixed lint 70e04dfc8686d02178eb0868d3bf78e3f7d54f4f

File: types/header.go (hunk 12 above)
Comment (0xAleksaOpacic) [reply types/header.go]: because I am using it for both UnmarshalJSON and MarshalJSON I didn't want to have it locally in both methods. What do you think of making it non-exportable

File: helper/hex/hex_test.go (hunk 7 above)
Comment (zivkovicmilos) [reply helper/hex/hex_test.go]: Hahah I didn't mean literally generate 10 random numbers and paste them here, I mean define a local function that generates however many random numbers you request, and use that array as the input for the test

cc @dbrajovic

File: types/header.go (hunk 10 above)
Comment (zivkovicmilos) [reply types/header.go]: I meant why is `RemoveHexPrefixFromByteArray` exported

File: types/header.go (hunk 12 above)
Comment (zivkovicmilos) [reply types/header.go]: It should not be exported outside of the scope you're using it in

File: helper/hex/hex_test.go (hunk 7 above)
Comment (0xAleksaOpacic) [reply helper/hex/hex_test.go]: What am I getting with that test, what random number will show me that something is not working, why would I have the necessary generation if it doesn't test anything?  
This subset of numbers contains min, max, and few in between which covers all test cases. 
I don't like implementing anything with a random generator in the test. I need to be sure that the test covers all cases if not then the test is not correct. 
If you guys want this we can add it. Don't see any value in adding a generator method locally. Maybe if we add a global util function that we can use from all unit tests it will make sense :)
cc @Kourin1996

File: types/header.go (hunk 10 above)
Comment (0xAleksaOpacic) [reply types/header.go]: I saw that we have pattern [2:] in another packages also in the future maybe i am going to need this somewhere outside of this package, that's why i left it exported

File: types/header.go (hunk 12 above)
Comment (0xAleksaOpacic) [reply types/header.go]: agreed 4816d822d636264251cac988f4c23f57c2841c34

File: types/header.go (hunk 11 above)
Comment (0xAleksaOpacic) [reply types/header.go:173]: Agree i can not be sure, but to be honest not sure how would i write this different, please add suggestion :)

File: helper/hex/hex_test.go (hunk 7 above)
Comment (Kourin1996) [reply helper/hex/hex_test.go]: I agree with @Aleksao998's opinion. In my mind, all inputs should be defined statically. Should think about random test separately.

File: types/header_test.go (hunk 4 above)
Comment (0xAleksaOpacic) [reply types/header_test.go]: fixed 24f329c560da5a8abf82615c553ae4508bd5925f. please check if i should add more test cases :)

Comment (0xAleksaOpacic) [reply types/header_test.go]: 24f329c560da5a8abf82615c553ae4508bd5925f

**Review by Kourin1996 — APPROVED** [review]
LGTM. No new comments from my side. Please resolve the discussions raised by @zivkovicmilos

**Comment by zivkovicmilos** [comment]
Hey @scalalang2.

Confirming that this PR doesn't fix your issue - it's probably a mismatch in RLP encoding between Edge and the `go-ethereum` client. We'll look into this in a separate PR 🙏

**Review by zivkovicmilos — APPROVED** [review]

File: helper/hex/hex_test.go [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+	}
+
+	for _, value := range generateRandomNumbers(100) {
+		encodedValue := EncodeUint64(value)
+
+		decodedValue, err := DecodeUint64(encodedValue)
+		assert.NoError(t, err)
+
+		assert.Equal(t, value, decodedValue)
+	}
```
Comment (Kourin1996) [inline helper/hex/hex_test.go]: Why do you need to test with random numbers...? Min, max, and mid possible numbers are enough
+
Encode and Decode in a test is a different way to test. You should focus on testing new code in new test

**Review by Kourin1996 — CHANGES_REQUESTED** [review]

File: helper/hex/hex_test.go (hunk 13 above)
Comment (zivkovicmilos) [reply helper/hex/hex_test.go]: We never hardcoded number values before in tests, nor should we - these numbers truly need to be random as we're not testing random number generation but encoding and decoding.

The functionality that was added as part of this PR (header encoding / decoding) is covered in another test

Comment (Kourin1996) [reply helper/hex/hex_test.go]: I don't know why we need to test with random value...

There are 2 things I want to say

1. What is the purpose of the test? Do we need to test whether go can calculate `1+1` correctly? Or course, we don't need this. Likewise, it's not important to test that `strings.TrimPrefix` and `strconv.ParseUint` work correctly. The important thing is to check the DecodeUint64 can parse all of the values in the range we expect and return the correct value. (Otherwise, this function returns error or not) In such a case, do we need to test with random value? In such a case, we should do boundary value test. (In addition, we should give the value with wrong format to check the function returns error) You said "We never hardcoded number values" but the basis of unit test is to give sample inputs and check whether the outputs equal to expected values.

2. Why does new test encode and decode? The new code is to decode only. We should add the test to check the function first. But this test does both encode and decode. We should thing about unit test, e2e test, and fuzzy test separately. The test that checks many things is hard to read and difficult to fix.
 
And I'm not talking about the test for header encoding / decoding. But this test encodes uint64 values and decodes it.

Comment (Kourin1996) [reply helper/hex/hex_test.go]: @zivkovicmilos

Comment (zivkovicmilos) [reply helper/hex/hex_test.go]: Got it, no problem in keeping it minimal, you're right that we double verify stuff here.

Fixed:
[fe561e7](https://github.com/0xPolygon/polygon-edge/pull/642/commits/fe561e7fee2ca2975e121541ea2d65ce5a1ddee3)

**Review by Kourin1996 — APPROVED** [review]
LGTM

File: helper/hex/hex_test.go (hunk 13 above)
Comment (Kourin1996) [reply helper/hex/hex_test.go]: Thank you for fixing!
